pennylane/gradients/spsa_gradient.py

```diff
@@ -0,0 +1,629 @@
+              the sampler.
+
+        sampler_seed (int or None): Seed passed to ``sampler``. The seed is passed in each
+            call to the sampler, so that only one unique direction is sampled even if
+            ``num_directions>1``.
+
+    Returns:
+        tensor_like or tuple[list[QuantumTape], function]:
+
+        - If the input is a QNode, a tensor
+          representing the output Jacobian matrix of size
+          ``(number_outputs, number_gate_parameters)`` is returned.
+
+        - If the input is a tape, a tuple containing a list of generated tapes,
+          in addition to a post-processing function to be applied to the
+          evaluated tapes.
+
+    **Example**
+
+    This gradient transform can be applied directly to :class:`QNode <pennylane.QNode>`
+    objects:
+
+    >>> @qml.qnode(dev)
+    ... def circuit(params):
+    ...     qml.RX(params[0], wires=0)
+    ...     qml.RY(params[1], wires=0)
+    ...     qml.RX(params[2], wires=0)
+    ...     return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(0))
+    >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+    >>> qml.gradients.spsa_grad(circuit)(params)
+    tensor([[-0.19280803, -0.19280803,  0.19280803],
+            [ 0.34786926,  0.34786926, -0.34786926]], requires_grad=True)
+
+    Note that the SPSA gradient is a statistical estimator that uses a given number of
+    function evaluations that does not depend on the number of parameters. While this
+    bounds the cost of the estimation, it also implies that the returned values are not
+    exact (even for devices with ``shots=None``) and that they will fluctuate.
+    See the usage details below for more information.
+
+    .. details::
+        :title: Usage Details
+
+        The number of directions in which the derivative is computed to estimate the gradient
+        can be controlled with the keyword argument ``num_directions``. For the QNode above,
+        a more precise gradient estimation from ``num_directions=20`` directions yields
+
+        >>> qml.gradients.spsa_grad(circuit, num_directions=20)(params)
+        tensor([[-0.32969058, -0.18924389, -0.28716881],
+                [ 0.59483632,  0.34143874,  0.51811744]], requires_grad=True)
+
+        We may compare this to the exact value obtained from parameter-shift rules:
+
+        >>> qml.gradients.param_shift(circuit)(params)
+        tensor([[-0.3875172 , -0.18884787, -0.38355704],
+                [ 0.69916862,  0.34072424,  0.69202359]], requires_grad=True)
+
+        As we can see, the SPSA output is a rather coarse approximation to the true
+        gradient, and this although the parameter-shift rule for three parameters uses
+        just six circuit evaluations, much fewer than SPSA! Consequentially, SPSA is
+        not necessarily useful for small circuits with few parameters, but will pay off
+        for large circuits where other gradient estimators require unfeasibly many circuit
+        executions.
```

Comment:
I'm not quite following what you're saying here. I thiiiink it's just the result of a typo but overall it sounds like you're saying parameter-shift is cheaper and more accurate, which kinda leaves me (aka the reader) wondering what the point SPSA is. I think the "although" is what is throwing me off 😄

Reply:
I rephrased this part, let me know what you think :)

Reply:
Thank you!

Reply:
looks great, ty!